core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -0,0 +1,394 @@
+            public void rollback() {
+                try {
+                    aRef.set(null);
+                    aSettingUpdater.rollback();
+                } finally {
+                    bRef.set(null);
+                    bSettingUpdater.rollback();
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "CompoundUpdater for: " + aSettingUpdater + " and " + bSettingUpdater;
+            }
+        };
+    }
+
+
+    private class Updater implements SettingsService.SettingUpdater {
+        private final Consumer<T> consumer;
+        private final ESLogger logger;
+        private final Predicate<T> accept;
+        private String value;
+        private boolean commitPending;
+        private String pendingValue;
+        private T valueInstance;
+
+        public Updater(Consumer<T> consumer, ESLogger logger, Settings settings, Predicate<T> accept) {
+            this.consumer = consumer;
+            this.logger = logger;
+            value = getRaw(settings);
+            this.accept = accept;
+        }
+
+
+        public boolean prepareApply(Settings settings) {
+            String newValue = settings.get(key);
+            if (newValue == null) {
+                newValue = getRaw(settings);
+            }
+            if (value.equals(newValue) == false) {
+                T inst = get(settings);
+                if (accept.test(inst) == false) {
+                    throw new IllegalArgumentException("illegal value can't update [" + key + "] from [" + value + "] to [" + getRaw(settings) + "]");
+                }
+                logger.info("update [{}] from [{}] to [{}]", key, value, getRaw(settings));
+                pendingValue = newValue;
+                valueInstance = inst;
+                commitPending = true;
+
+            } else {
+                commitPending = false;
+            }
+            return commitPending;
+        }
+
+        public void apply() {
+            if (commitPending) {
+                value = pendingValue;
+                consumer.accept(valueInstance);
+            }
+            commitPending = false;
+            valueInstance = null;
+            pendingValue = null;
+        }
+
+        public void rollback() {
+            commitPending = false;
+            valueInstance = null;
+            pendingValue = null;
+        }
+
+        @Override
+        public String toString() {
+            return "Updater for: " + Setting.this.toString();
+        }
+    }
+
+
+    public Setting(String key, String description, String defaultValue, Function<String, T> parser, boolean dynamic, Scope scope) {
+        this(key, description, (s) -> defaultValue, parser, dynamic, scope);
+    }
+
+    public static Setting<Float> floatSetting(String key, float defaultValue, boolean dynamic, Scope scope) {
+        return new Setting<>(key, "_na_", (s) -> Float.toString(defaultValue), Float::parseFloat, dynamic, scope);
```

Comment:
I dunno if we should allow settings with `_na_` as the description. Maybe we do that in another commit though.